### PR TITLE
Scheduler.process: apply directly to program unit corresponding to Item

### PR DESCRIPTION
**Pull request by reuterbal**
This is part 2 to #136 (and filed against that PR branch to allow consolidating the fix if acceptable).

The extended fix in this PR is more invasive than the minimum workaround implemented in #136 by applying transformations directly to the program unit corresponding to an item instead of relying on recursion to eventually end up at the right object. This also removes the need for the check of `item.name == routine.name` before applying transformations, because now we do actually apply transformations only directly to the subroutines that are meant to be transformed.

This highlighted also a flaw in the current scheduler design which effectively preempts processing contained member routines entirely, because we can't represent contained member routines as scheduler graph nodes (the discovery mechanism is unable to find member routines), and the above mentioned name check preempts recursion into the member routine. The implementation in this PR partially resolves this by getting rid of the name check. Proper representation of contained member routines requires more changes, though, and I will create a separate issue for that.

As this change is more invasive in the scheduler, it may have unintended side-effects and therefore this fix is optional.

### CODE REVIEW

**Comment by reuterbal** [comment]
Waiting for CI to confirm but I think I fixed the issue that was flagged by the ecwam regression test: The DependencyTransformation had an implicit dependency on the recursion behaviour of the Transformation class - both in terms of requiring the recursion to always start from the source file level, and trusting that recursion never actually reaches contained member procedures...

I added some safe guards to the transformation that ensure existing behaviour is retained. 

@mlange05: Please let me know if you would like to review this as a diff to #136, or want me to merge this PR into the other and review them together?

**Review by mlange05 — APPROVED** [review]
Ok, I've had a first look, and I think the final review should be on the combined PR for both. So I'm approving this now, but will take another look at the full thing.

I've already flagged a few things that strike me as odd in the dependency transform. This could very well be me not understanding some subtle detail though, so I'll need to play around with this myself a bit more. 

Reading through this, I'm also not quite convinced anymore that we even need the `recurse_to_contained` flag. I think the only use for this were member subroutines and the `sourcefile->module->subroutine` cascade which we're trying to suprpess with this, no? For the member-routine recurision (a rare case, I guess), an explicit `recurse_to_member_routines` or even a user-level recursive call to `transform_subroutine` might actually be better, no? 

Anyway let's merge this into #136 and see if we can whittle this down some more from there.

File: loki/bulk/scheduler.py
Comment (reuterbal) [reply loki/bulk/scheduler.py:656]: Indeed

File: loki/transform/dependency_transform.py
Comment (reuterbal) [reply loki/transform/dependency_transform.py:88]: It's due to the fact that the cleanest way of calling this transformation is via the file graph and then using the transformation's recursion to trickle down into the various nodes. However, with that the `item` property isn't updated cleanly for individual program units within the source file. This leads to outdated `targets` and, in a situation where driver and kernel would sit in the same module, we would assume the role of one of them for both.

**Comment by reuterbal** [comment]
As requested, merged into #136 for further review.

**Pull request merged**
